### PR TITLE
THRIFT-5553: java library to use newer gradle API

### PR DESCRIPTION
**Pull request by Jimexist**
java library to use newer gradle API, fixing most deprecation warning generated by `./gradlew help --scan`

- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [x] If your change does not involve any code, include `[skip ci]` anywhere in the commit message to free up build resources.

### CODE REVIEW

File: lib/java/gradle/environment.gradle [hunk 1]
```diff
@@ -64,14 +64,14 @@ repositories {
...
+    implementation "javax.servlet:javax.servlet-api:${servletVersion}"
+    implementation "javax.annotation:javax.annotation-api:${javaxAnnotationVersion}"
+    implementation "org.apache.commons:commons-lang3:3.12.0"
 
-    testCompile "junit:junit:${junitVersion}"
-    testCompile "org.mockito:mockito-all:${mockitoVersion}"
-    testRuntime "org.slf4j:slf4j-log4j12:${slf4jVersion}"
+    testImplementation "junit:junit:${junitVersion}"
+    testImplementation "org.mockito:mockito-all:${mockitoVersion}"
+    testRuntimeOnly "org.slf4j:slf4j-log4j12:${slf4jVersion}"
```
Comment (Jimexist) [inline lib/java/gradle/environment.gradle:76]:
```suggestion
    testRuntime"org.slf4j:slf4j-log4j12:${slf4jVersion}"
```

File: lib/java/gradle/functionalTests.gradle [hunk 2]
```diff
@@ -59,18 +59,18 @@ shadowJar {
...
     // We do not need a version number for this internal jar
-    version = null
+    archiveVersion = null
 
     // Bundle the complete set of unit test classes including generated code
     // and the runtime dependencies in one JAR to expedite execution.
     from sourceSets.test.output
     from sourceSets.crossTest.output
-    configurations = [project.configurations.testRuntime]
+    configurations = [project.configurations.testRuntimeOnly]
```
Comment (Jimexist) [inline lib/java/gradle/functionalTests.gradle:73]:
```suggestion
    configurations = [project.configurations.testRuntime]
```

File: lib/java/gradle/publishing.gradle [hunk 3]
```diff
@@ -46,7 +46,7 @@ task copyDependencies(type: Copy, group: 'Build') {
     project.assemble.dependsOn it
 
     destinationDir = file("$buildDir/deps")
-    from configurations.testRuntime
+    from configurations.testRuntimeOnly
```
Comment (Jimexist) [inline lib/java/gradle/publishing.gradle:49]:
```suggestion
    from configurations.testRuntime
```

**Review by ctubbsii — APPROVED** [review]
I don't know gradle, but I don't see anything particularly problematic in this PR.

**Comment by Jimexist** [comment]
i'm not sure i can debug the CI failures - seems an issue with travis itself as the failures do not have logs and they occur randomly (same situation with other pull requests)

**Comment by ctubbsii** [comment]
> i'm not sure i can debug the CI failures - seems an issue with travis itself as the failures do not have logs and they occur randomly (same situation with other pull requests)

I don't think those failures are related to this change, but I can't be sure. The individual tasks aren't well named and I'm not sure what they're doing.

Also, somebody familiar with the Travis CI configuration might want to investigate moving over to GitHub Actions, since 1) it's built-in to GitHub and possibly a bit more reliable/less dependencies on another party, and 2) I don't trust Travis CI's new permissions model that seems to require granting it oauth permissions for private repos in order to log in to Travis-CI.com to manage builds, even if you only want to use it with public repos. I wonder if that was partially the reason for [this recent security incident involving Travis CI app being targeted and used as a mechanism to steal private repo contents](https://github.blog/2022-04-15-security-alert-stolen-oauth-user-tokens/)

I know one of the main limits of GitHub Actions is the inability to use ARM or other non-x86 architectures when building. As a Java developer I don't generally care about that, but I'm guessing the Thrift project probably cares.

**Comment by Jimexist** [comment]
> > i'm not sure i can debug the CI failures - seems an issue with travis itself as the failures do not have logs and they occur randomly (same situation with other pull requests)
> 
> I don't think those failures are related to this change, but I can't be sure. The individual tasks aren't well named and I'm not sure what they're doing.
> 
> Also, somebody familiar with the Travis CI configuration might want to investigate moving over to GitHub Actions, since 1) it's built-in to GitHub and possibly a bit more reliable/less dependencies on another party, and 2) I don't trust Travis CI's new permissions model that seems to require granting it oauth permissions for private repos in order to log in to Travis-CI.com to manage builds, even if you only want to use it with public repos. I wonder if that was partially the reason for [this recent security incident involving Travis CI app being targeted and used as a mechanism to steal private repo contents](https://github.blog/2022-04-15-security-alert-stolen-oauth-user-tokens/)
> 
> I know one of the main limits of GitHub Actions is the inability to use ARM or other non-x86 architectures when building. As a Java developer I don't generally care about that, but I'm guessing the Thrift project probably cares.

I agree with that GitHub Actions more user friendly and efficient. If that's a direction that the community wishes to move towards I'd be able to help.

**Comment by ctubbsii** [comment]
> I agree with that GitHub Actions more user friendly and efficient. If that's a direction that the community wishes to move towards I'd be able to help.

That would be great. I think they could replace AppVeyor as well.

**Comment by Jimexist** [comment]
> > I agree with that GitHub Actions more user friendly and efficient. If that's a direction that the community wishes to move towards I'd be able to help.
> 
> That would be great. I think they could replace AppVeyor as well.

Let me create https://issues.apache.org/jira/browse/THRIFT-5564 and wait for some more feedback if any.

**Comment by Jimexist** [comment]
FYI i also had to update tutorial/js and tutorial/java since their ant build refers to the legacy java libs

**Comment by ctubbsii** [comment]
@Jimexist I merged master into this one to bring it up to date. I ran a `gradle build` and it seems to work fine with Gradle 6.9.2. However, it still shows a warning that `Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.`

**Comment by Jimexist** [comment]
> However, it still shows a warning that `Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.`

yes because we still have one thing left: migrate `maven` plugin to `maven-publish` plugin. that's a bigger task to handle so i'll do it later, along with upgrading to gradle 7.x

**Review by ctubbsii — APPROVED** [review]
Looks fine to me. Will wait for CI. In the meantime, @Jimexist , please make sure that there wasn't any errors introduced while resolving merge conflicts, and that this is still ready to go.

**Comment by Jimexist** [comment]
> Looks fine to me. Will wait for CI. In the meantime, @Jimexist , please make sure that there wasn't any errors introduced while resolving merge conflicts, and that this is still ready to go.

thanks! confirmed on my side that this is 👍

**Comment by Jimexist** [comment]
> > However, it still shows a warning that `Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.`
> 
> yes because we still have one thing left: migrate `maven` plugin to `maven-publish` plugin. that's a bigger task to handle so i'll do it later, along with upgrading to gradle 7.x

I've put up:
- #2601 to migrate to gradle 7.4+ which is the latest version

**Pull request merged**